### PR TITLE
feat: add StampSettingsInfo and STAMP_ATTRS

### PR DESCRIPTION
**Pull request by gregmagolan**
Adds support for stamp env vars `BAZEL_VERSION_FILE` and `BAZEL_INFO_FILE` to `run_node` if `"stamp": STAMP_ATTR` is in `ctx.attr`

Adds `stamp` attr to `npm_package_bin`
Adds `stamp` attr to `pkg_npm` (replacing `node_context_info`)
Adds `stamp` attr to `pkg_web` (replacing `node_context_info`)
Adds `stamp` attr to `esbuild` (replacing `node_context_info`)
Adds `stamp` attr to `rollup_bundle` (replacing `node_context_info`)

Removes old NodeContextData and NODE_CONTEXTS_ATTRS from `//internal/node/context.bzl` in favor of StampSettingInfo and STAMP_ATTR from `//nodejs/private/providers:stamp_setting_info.bzl`

### CODE REVIEW

**Comment by Wenqer** [comment]
Is it possible to add 'stamp' attr to nodejs_binary and nodejs_test too?

**Comment by gregmagolan** [comment]
> Is it possible to add 'stamp' attr to nodejs_binary and nodejs_test too?

Definitely. @alexeagle or I will add that as a follow-up to this.

File: internal/providers/node_runtime_deps_info.bzl [hunk 1]
```diff
@@ -141,6 +142,15 @@ def run_node(ctx, inputs, arguments, executable, chdir = None, **kwargs):
         bazel_node_module_roots = bazel_node_module_roots + "%s:%s" % (path, root)
     env["BAZEL_NODE_MODULES_ROOTS"] = bazel_node_module_roots
 
+    stamp = ctx.attr.stamp[StampSettingInfo].value if hasattr(ctx.attr, "stamp") else False
+    if stamp:
+        env["BAZEL_VERSION_FILE"] = ctx.version_file.path
+        env["BAZEL_INFO_FILE"] = ctx.info_file.path
+        if type(inputs) == "list":
```
[the remark this answers is not in the document]
Comment (gregmagolan) [reply internal/providers/node_runtime_deps_info.bzl:149]: 😎

File: internal/pkg_web/test/BUILD.bazel [hunk 2]
```diff
@@ -79,7 +73,7 @@ pkg_web(
         ":bundle.min_debug",
         ":bundle.min_debug.es2015",
     ],
-    node_context_data = ":force_stamp",
+    stamp = "@build_bazel_rules_nodejs//nodejs:always_stamp",
```
[the remark this answers is not in the document]
Comment (gregmagolan) [reply internal/pkg_web/test/BUILD.bazel]: Done

File: providers.bzl [hunk 3]
```diff
@@ -51,13 +51,30 @@ load(
...
-    _NodeContextInfo = "NodeContextInfo",
+    "//nodejs/private/providers:stamp_setting_info.bzl",
+    _StampSettingInfo = "StampSettingInfo",
 )
 
-NodeContextInfo = _NodeContextInfo
-NODE_CONTEXT_ATTRS = _NODE_CONTEXT_ATTRS
+# TODO(5.0): Remove this. Temporarily here to land
+# https://github.com/bazelbuild/rules_nodejs/pull/3162 on 5.x branch but a
+# follow up PR should remove this and switch ALL refences to //nodejs to use
```
[the remark this answers is not in the document]
Comment (gregmagolan) [reply providers.bzl]: Done

File: nodejs/BUILD.bazel [hunk 4]
```diff
@@ -49,3 +50,32 @@ user_args(
...
+    name = "use_stamp_flag",
+    stamp = select({
+        "//nodejs:stamp": True,
+        "//conditions:default": False,
+    }),
+    visibility = ["//visibility:public"],
+)
+
+stamp_setting(
+    name = "always_stamp",
```
[the remark this answers is not in the document]
Comment (gregmagolan) [reply nodejs/BUILD.bazel]: Done

**Review by alexeagle — APPROVED** [review]
Delightful :octocat:

**Comment by mtsgrd** [comment]
> > Is it possible to add 'stamp' attr to nodejs_binary and nodejs_test too?
> 
> Definitely. @alexeagle or I will add that as a follow-up to this.

Is there an issue tracking this? I would currently use it to stamp a cypress e2e testing API key into the environment variables for `cypress_web_test`.

**Pull request merged**